File: src/heavyindustry/world/blocks/distribution/TubeConveyor.java
package heavyindustry.world.blocks.distribution;

import arc.graphics.g2d.*;
import arc.math.geom.*;
import arc.util.*;
import mindustry.entities.units.*;
import mindustry.gen.*;
import mindustry.graphics.*;
import mindustry.type.*;
import mindustry.world.blocks.distribution.*;

import static arc.Core.*;
import static heavyindustry.util.Utils.*;
import static mindustry.Vars.*;

/**
 * Compared to CoveredConverter, its upper layer texture has been changed to one that can have light and shadow effects.
 *
 * @author Eipusino
 */
public class TubeConveyor extends BeltConveyor {
    public static final float itemSpace = 0.4f;
    public static final int[][] tiles = new int[][]{
            {},
            {0, 2}, {1, 3}, {0, 1},
            {0, 2}, {0, 2}, {1, 2},
            {0, 1, 2}, {1, 3}, {0, 3},
            {1, 3}, {0, 1, 3}, {2, 3},
            {0, 2, 3}, {1, 2, 3}, {0, 1, 2, 3}
    };

    public TextureRegion[][] topRegion;
    public TextureRegion[] capRegion;
    public TextureRegion editorRegion;

    public TubeConveyor(String name) {
        super(name);
    }

    @Override
    public void load() {
        super.load();
        topRegion = splitLayers(name + "-sheet", 32, 2);
        capRegion = new TextureRegion[]{topRegion[1][0], topRegion[1][1]};
        editorRegion = atlas.find(name + "-editor");
    }

    @Override
    public TextureRegion[] icons() {
        return new TextureRegion[]{editorRegion};
    }

    @Override
    public void drawPlanRegion(BuildPlan plan, Eachable<BuildPlan> list) {
        int[] bits = getTiling(plan, list);

        if (bits == null) return;

        TextureRegion conveyor = conveyorAtlas[0][bits[0]];
        Draw.rect(conveyor, plan.drawx(), plan.drawy(), conveyor.width * bits[1] * conveyor.scl(), conveyor.height * bits[2] * conveyor.scl(), plan.rotation * 90);

        BuildPlan[] directionals = new BuildPlan[4];
        list.each(other -> {
            if (other.breaking || other == plan) return;

            int i = 0;
            for (Point2 point : Geometry.d4) {
                int x = plan.x + point.x, y = plan.y + point.y;
                if (x >= other.x - (other.block.size - 1) / 2 && x <= other.x + (other.block.size / 2) && y >= other.y - (other.block.size - 1) / 2 && y <= other.y + (other.block.size / 2)) {
                    if ((other.block instanceof Conveyor ? (plan.rotation == i || (other.rotation + 2) % 4 == i) : !noSideBlend && ((plan.rotation == i && other.block.acceptsItems) || (plan.rotation != i && other.block.outputsItems())))) {
                        directionals[i] = other;
                    }
                }
                i++;
            }
        });

        int mask = 0;
        for (int i = 0; i < directionals.length; i++) {
            if (directionals[i] != null) {
                mask += (1 << i);
            }
        }
        mask |= (1 << plan.rotation);
        Draw.rect(topRegion[0][mask], plan.drawx(), plan.drawy(), 0);
        for (int i : tiles[mask]) {
            if (directionals[i] == null || (directionals[i].block instanceof Conveyor ? (directionals[i].rotation + 2) % 4 == plan.rotation : ((plan.rotation == i && !directionals[i].block.acceptsItems) || (plan.rotation != i && !directionals[i].block.outputsItems())))) {
                int id = i == 0 || i == 3 ? 1 : 0;
                Draw.rect(capRegion[id], plan.drawx(), plan.drawy(), i == 0 || i == 2 ? 0 : -90);
            }
        }
    }

    public class TubeConveyorBuild extends BeltConveyorBuild {
        public int tiling = 0;

        public Building buildAt(int i) {
            return nearby(i);
        }

        public boolean valid(int i) {
            Building b = buildAt(i);
            return b != null && (b instanceof TubeConveyorBuild ? (b.front() != null && b.front() == this) : b.block.acceptsItems || b.block.outputsItems());
        }

        public boolean isEnd(int i) {
            Building b = buildAt(i);
            return (!valid(i) && (b == null ? null : b.block) != this.block) || (b instanceof ConveyorBuild && ((b.rotation + 2) % 4 == rotation || (b.front() != this && back() == b)));
        }

        @Override
        public void draw() {
            int frame = enabled && clogHeat <= 0.5f ? (int) (((Time.time * speed * 8f * timeScale * efficiency)) % 4) : 0;

            //draw extra conveyors facing this one for non-square tiling purposes
            Draw.z(Layer.blockUnder);
            for (int i = 0; i < 4; i++) {
                if ((blending & (1 << i)) != 0) {
                    int dir = rotation - i;
                    float rot = i == 0 ? rotation * 90 : (dir) * 90;

                    Draw.rect(sliced(conveyorAtlas[frame][0], i != 0 ? SliceMode.bottom : SliceMode.top), x + Geometry.d4x(dir) * tilesize * 0.75f, y + Geometry.d4y(dir) * tilesize * 0.75f, rot);
                    Draw.rect(sliced(topRegion[0][1], i != 0 ? SliceMode.bottom : SliceMode.top), x + Geometry.d4x(dir) * tilesize * 0.75f, y + Geometry.d4y(dir) * tilesize * 0.75f, rot);
                }
            }

            Draw.z(Layer.block - 0.25f);

            Draw.rect(conveyorAtlas[frame][blendbits], x, y, tilesize * blendsclx, tilesize * blendscly, rotation * 90);

            Draw.z(Layer.block - 0.2f);
            float layer = Layer.block - 0.2f, wwidth = world.unitWidth(), wheight = world.unitHeight(), scaling = 0.01f;

            for (int i = 0; i < len; i++) {
                Item item = ids[i];
                Tmp.v1.trns(rotation * 90, tilesize, 0);
                Tmp.v2.trns(rotation * 90, -tilesize / 2f, xs[i] * tilesize / 2f);

                float ix = (x + Tmp.v1.x * ys[i] + Tmp.v2.x), iy = (y + Tmp.v1.y * ys[i] + Tmp.v2.y);

                //keep draw position deterministic.
                Draw.z(layer + (ix / wwidth + iy / wheight) * scaling);
                Draw.rect(item.fullIcon, ix, iy, itemSize, itemSize);
            }

            Draw.z(Layer.block - 0.15f);
            Draw.rect(topRegion[0][tiling], x, y, 0);
            int[] placementID = tiles[tiling];
            for (int i : placementID) {
                if (isEnd(i)) {
                    int id = i == 0 || i == 3 ? 1 : 0;
                    Draw.rect(capRegion[id], x, y, i == 0 || i == 2 ? 0 : -90);
                }
            }
        }

        @Override
        public int acceptStack(Item item, int amount, Teamc source) {
            if (isEnd(reverse(rotation)) && items.total() >= 2) return 0;
            if (isEnd(reverse(rotation)) && isEnd(rotation) && items.total() >= 1) return 0;
            return Math.min((int) (minitem / itemSpace), amount);
        }

        @Override
        public void drawCracks() {
            Draw.z(Layer.block);
            super.drawCracks();
        }

        @Override
        public void unitOn(Unit unit) {
            //There is a cover, can't slide on this thing.
        }

        @Override
        public void onProximityUpdate() {
            super.onProximityUpdate();

            tiling = 0;
            for (int i = 0; i < 4; i++) {
                Building otherBlock = nearby(i);
                if (otherBlock == null) continue;
                if ((otherBlock.block instanceof Conveyor ? (rotation == i || (otherBlock.rotation + 2) % 4 == i) : !noSideBlend && ((rotation == i && otherBlock.block.acceptsItems) || (rotation != i && otherBlock.block.outputsItems())))) {
                    tiling |= (1 << i);
                }
            }
            tiling |= 1 << rotation;
        }
    }
}

File: src/heavyindustry/world/blocks/distribution/TubeConveyor.kt
package heavyindustry.world.blocks.distribution

import heavyindustry.util.HIUtils
import arc.Core
import arc.graphics.g2d.Draw
import arc.graphics.g2d.TextureRegion
import arc.math.geom.Geometry
import arc.util.Eachable
import arc.util.Time
import arc.util.Tmp
import mindustry.entities.units.BuildPlan
import mindustry.gen.Building
import mindustry.gen.Teamc
import mindustry.gen.Unit
import mindustry.graphics.Layer
import mindustry.type.Item
import mindustry.world.blocks.Autotiler.SliceMode
import mindustry.world.blocks.distribution.Conveyor
import kotlin.math.min

import mindustry.Vars.itemSize
import mindustry.Vars.tilesize
import mindustry.Vars.world

/**
 * Compared to CoveredConverter, its upper layer texture has been changed to one that can have light and shadow effects.
 */
open class TubeConveyor(name: String) : BeltConveyor(name) {
    companion object {
        val itemSpace = 0.4f
        val tiles: Array<IntArray> = arrayOf(
            intArrayOf(),
            intArrayOf(0, 2), intArrayOf(1, 3), intArrayOf(0, 1),
            intArrayOf(0, 2), intArrayOf(0, 2), intArrayOf(1, 2),
            intArrayOf(0, 1, 2), intArrayOf(1, 3), intArrayOf(0, 3),
            intArrayOf(1, 3), intArrayOf(0, 1, 3), intArrayOf(2, 3),
            intArrayOf(0, 2, 3), intArrayOf(1, 2, 3), intArrayOf(0, 1, 2, 3)
        )
    }

    lateinit var topRegion: Array<Array<TextureRegion>>
    lateinit var capRegion: Array<TextureRegion>

    override fun drawPlanRegion(plan: BuildPlan, list: Eachable<BuildPlan>) {
        super.drawPlanRegion(plan, list)
        val directionals = arrayOfNulls<BuildPlan>(4)
        list.each { other: BuildPlan ->
            if (other.breaking || other === plan) return@each
            for ((i, point) in Geometry.d4.withIndex()) {
                val x = plan.x + point.x
                val y = plan.y + point.y
                if ((x >= other.x - (other.block.size - 1) / 2 && x <= other.x + (other.block.size / 2)) && y >= other.y - (other.block.size - 1) / 2 && y <= other.y + (other.block.size / 2)) {
                    if ((if (other.block is Conveyor) (plan.rotation == i || (other.rotation + 2) % 4 == i) else ((plan.rotation == i && other.block.acceptsItems) || (plan.rotation != i && other.block.outputsItems())))) {
                        directionals[i] = other
                    }
                }
            }
        }

        var mask = 0
        for (i in directionals.indices) {
            if (directionals[i] != null) {
                mask += (1 shl i)
            }
        }
        mask = mask or (1 shl plan.rotation)
        Draw.rect(topRegion[0][mask], plan.drawx(), plan.drawy(), 0f)
        for (i in tiles[mask]) {
            if (directionals[i] == null || (if (directionals[i]!!.block is Conveyor) (directionals[i]!!.rotation + 2) % 4 == plan.rotation else ((plan.rotation == i && !directionals[i]!!.block.acceptsItems) || (plan.rotation != i && !directionals[i]!!.block.outputsItems())))) {
                val id = if (i == 0 || i == 3) 1 else 0
                Draw.rect(capRegion[id], plan.drawx(), plan.drawy(), (if (i == 0 || i == 2) 0 else -90).toFloat())
            }
        }
    }

    override fun load() {
        super.load()
        topRegion = HIUtils.splitLayers("$name-sheet", 32, 2)
        capRegion = arrayOf(topRegion[1][0], topRegion[1][1])
    }

    override fun icons(): Array<TextureRegion> {
        return arrayOf(Core.atlas.find("$name-icon-editor"))
    }

    open inner class TubeConveyorBuild : BeltConveyorBuild() {
        var tiling: Int = 0
        var calls: Int = 0
        override fun updateProximity() {
            super.updateProximity()
            calls++
        }

        fun buildAt(i: Int): Building? {
            return nearby(i)
        }

        fun valid(i: Int): Boolean {
            val b = buildAt(i)
            return b != null && (if (b is TubeConveyorBuild) (b.front() != null && b.front() === this) else (b.block.acceptsItems || b.block.hasItems))
        }

        fun isEnd(i: Int): Boolean {
            val b = buildAt(i)
            return (!valid(i) && b?.block !== this.block) || (b is ConveyorBuild && ((b.rotation + 2) % 4 == rotation || (b.front() !== this && back() === b)))
        }

        override fun draw() {
            val frame = if (enabled && clogHeat <= 0.5f) (((Time.time * speed * 8f * timeScale * efficiency)) % 4).toInt() else 0

            //draw extra conveyors facing this one for non-square tiling purposes
            Draw.z(Layer.blockUnder)
            for (i in 0..3) {
                if ((blending and (1 shl i)) != 0) {
                    val dir = rotation - i
                    val rot = (if (i == 0) rotation * 90 else (dir) * 90).toFloat()

                    Draw.rect(sliced(regions[0][frame], if (i != 0) SliceMode.bottom else SliceMode.top), x + Geometry.d4x(dir) * tilesize * 0.75f, y + Geometry.d4y(dir) * tilesize * 0.75f, rot)
                }
            }

            Draw.z(Layer.block - 0.25f)

            Draw.rect(conveyorParts[frame][blendbits], x, y, (tilesize * blendsclx).toFloat(), (tilesize * blendscly).toFloat(), (rotation * 90).toFloat())

            Draw.z(Layer.block - 0.2f)
            val layer = Layer.block - 0.2f
            val wwidth = world.unitWidth().toFloat()
            val wheight = world.unitHeight().toFloat()
            val scaling = 0.01f

            for (i in 0 until len) {
                val item = ids[i]
                Tmp.v1.trns((rotation * 90).toFloat(), tilesize.toFloat(), 0f)
                Tmp.v2.trns((rotation * 90).toFloat(), -tilesize / 2f, xs[i] * tilesize / 2f)

                val ix = (x + Tmp.v1.x * ys[i] + Tmp.v2.x)
                val iy = (y + Tmp.v1.y * ys[i] + Tmp.v2.y)

                //keep draw position deterministic.
                Draw.z(layer + (ix / wwidth + iy / wheight) * scaling)
                Draw.rect(item.fullIcon, ix, iy, itemSize, itemSize)
            }

            Draw.z(Layer.block - 0.15f)
            Draw.rect(topRegion[0][tiling], x, y, 0f)
            val placementID = tiles[tiling]
            for (i in placementID) {
                if (isEnd(i)) {
                    val id = if (i == 0 || i == 3) 1 else 0
                    Draw.rect(capRegion[id], x, y, (if (i == 0 || i == 2) 0 else -90).toFloat())
                }
            }
        }

        override fun acceptStack(item: Item, amount: Int, source: Teamc): Int {
            if (isEnd(HIUtils.reverse(rotation)) && items.total() >= 2) return 0
            if (isEnd(HIUtils.reverse(rotation)) && isEnd(rotation) && items.total() >= 1) return 0
            return min((minitem / itemSpace).toInt().toDouble(), amount.toDouble()).toInt()
        }

        override fun unitOn(unit: Unit) {
        }

        override fun drawCracks() {
            Draw.z(Layer.block)
            super.drawCracks()
        }

        override fun onProximityUpdate() {
            super.onProximityUpdate()
            noSleep()
            next = front()
            nextc = if (next is TubeConveyorBuild && next.team === team) next as TubeConveyorBuild else null
            aligned = nextc != null && rotation == next.rotation

            tiling = 0
            for (i in 0..3) {
                val otherblock = nearby(i) ?: continue
                if ((if (otherblock.block is Conveyor) (rotation == i || (otherblock.rotation + 2) % 4 == i) else ((rotation == i && otherblock.block.acceptsItems) || (rotation != i && otherblock.block.outputsItems())))) {
                    tiling = tiling or (1 shl i)
                }
            }
            tiling = tiling or (1 shl rotation)
        }
    }
}
